fix(android): fix for changing tabs properties after creation

ypbnv opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27075

**Description:**
Fix for changing TabGroup's tabs' properties after they are created. Now every implementation of
AbstractTabGroup must take care of the exposed properties. Improvement of properties handling - they
are now separated for title, tilteColor, background drawable and icon. Imports clean up.

**Note:** I haven't updated the unit-tests for the following reason - currently our unit tests expect Tab instances to be tested without being added in a TabGroup. In this case I can't connect properties of the TabProxy with their counterparts in the native implementation - TabLayout.Tabs or MenuItems, because they are not created yet. That got me thinking "Do we want to have the native components be created without them being used in an actual drawn TabGroup?". If we do - this will need some rework of the code there. I checked what was the case in 7.5.0.GA (before the refactoring) and it does not create the native tabs if they are not added in a TabGroup (and they can't be added to another type of container directly) What do you think?

**Test case:**
There is a good sample code in the JIRA ticket.

ypbnv committed:
fix(android): fix for changing tabs properties after creation

Fix for changing TabGroup's tabs' properties after they are created. Now every implementation of
AbstractTabGroup must take care of the exposed properties. Improvement of properties handling - they
are now separated for title, tilteColor, background drawable and icon. Imports clean up.

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  2 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10950/1/artifact/dist/mobilesdk-8.2.0.v20190607045209-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 1 tests have failed
There are 1 tests failing and 464 skipped out of 3613 total tests.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.Titanium.Network.HTTPClient</td><td>clearCookieUnaffectedCheck</td><td>60.008</td><td><pre>Error: timeout of 60000ms exceeded
at Titanium.&lt;anonymous&gt; (/ti-mocha.js:4326:18)</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e1ec5e29470aef3d7683d37ece7e36622b25e6e0
</p>

jquick-axway commented:
@ypbnv, would you mind testing your PR with my "Don't keep activities" test shown in the ticket below please?
https://jira.appcelerator.org/browse/TIMOB-26964

In this test, when you display a child activity window from a `TabGroup`, the Android OS will temporarily destroy the `TabGroup` activity and all of its view/fragments, but the proxies will still be kept alive. When you navigate back, the Android OS will **restore** the `TabGroup` activity and the proxies will re-create the UI. We want to make sure this behavior still works. Thanks.

ypbnv commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIBottomNavigationTabGroup.java:
```diff
@@ -264,6 +233,65 @@ public void setBackgroundColor(int colorInt)
 	}
 
+	@Override
+	public void updateTabBackgroundDrawable(int index)
```
It is ignoring the index because we use the method whenever a new item is added in the tab controller and BottomNavigationMenuView is rebuilding the whole view when this happens. Thus everything not set through the "proper" APIs is reset. This is why we ignore it when setting the background drawable and the text color state list - the component does not provide a direct way to set those properties so I am relying on going through the layout and manually set the new values.

But it will be better to stick to going through all the items only when a new item is added, so I will move this logic in a new private method dealing with new items and respect the index value for other updates.

jquick-axway replied:
Okay. In this case, this is where a code comment would be helpful. Thanks.

jquick-axway replied:
Oh never mind. I can see you've already written a code comment about this.

ypbnv commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUITab.java:
```diff
@@ -22,6 +22,7 @@
 	public TiUITab(TabProxy proxy)
 	{
 		super(proxy);
+		proxy.setModelListener(this);
```
It is a leftover from another way I tried to implement, that did not work out as I wanted.

ypbnv commented:
@jquick-axway I ran the test for recreating the TabGroup after it being destroyed and it looks fine - after closing the child window the tabs are there and the TabGroup works just like before.

ypbnv committed:
fix(android): Go through all bottom style tabs only when necessary

Set the background drawalbe and title color state list only when a new item is added. Then the whole
menu is recreated and some of the properties may get wiped. In other updates use the index property
for updates. Remove one unnecessary line.

jquick-axway reviewed (approved):
CR: Pass

ssjsamir committed:
Merge branch 'master' into TIMOB-27075

ssjsamir reviewed (approved):
**FR Passed:** Able to change tab properties after creation, tested using the following test case: 
```
var win1 = Ti.UI.createWindow({

backgroundColor: 'blue',
title: 'Blue'
});
var button = Ti.UI.createButton({ title: 'Change tab' });
button.addEventListener('click', function () {
tab1.icon = '/images/ChatBadge.png';
tab1.title = 'changed';
});
win1.add(button);
 
var win2 = Ti.UI.createWindow({
backgroundColor: 'red',
title: 'Red'
});
win2.add(Ti.UI.createLabel({ text: 'I am a red window.' }));
 
var tab1 = Ti.UI.createTab({
window: win1,
title: 'Blue'
}),
tab2 = Ti.UI.createTab({
window: win2,
title: 'Red'
}),
tabGroup = Ti.UI.createTabGroup({
tabs: [ tab1, tab2 ]
});
tabGroup.open();
```

**Test Environment**
```
Pixel XL (7.1)
7.0.12-70X.4
Operating System Name: Mac OS Mojave
Operating System Version: 10.14.2
Node.js Version: 8.11.3
```

lokeshchdhry committed:
Merge branch 'master' into TIMOB-27075
